chore: Add unit tests for ToolboxSyncClient

anubhav756 opened the pull request:
```
Name                              Stmts   Miss  Cover
-----------------------------------------------------
src/toolbox_core/__init__.py          3      0   100%
src/toolbox_core/client.py          102      0   100%
src/toolbox_core/protocol.py         32      0   100%
src/toolbox_core/sync_client.py      41      0   100%
src/toolbox_core/sync_tool.py        38      0   100%
src/toolbox_core/tool.py             95      0   100%
src/toolbox_core/utils.py            37      0   100%
-----------------------------------------------------
```

kurtisvg commented on packages/toolbox-core/tests/test_sync_client.py:
```diff
@@ -0,0 +1,612 @@
+TEST_BASE_URL = "http://toolbox.example.com"
+
+
+@pytest.fixture(autouse=True)
+def manage_sync_client_class_state():
+    """
+    Resets the class-level event loop and thread for ToolboxSyncClient
+    before and after each test to ensure test isolation.
+    """
+    # Save current state if any
+    original_loop = getattr(ToolboxSyncClient, "_ToolboxSyncClient__loop", None)
+    original_thread = getattr(ToolboxSyncClient, "_ToolboxSyncClient__thread", None)
+
+    # Reset for the test
+    ToolboxSyncClient._ToolboxSyncClient__loop = None
+    ToolboxSyncClient._ToolboxSyncClient__thread = None
+
+    yield  # Run the test
+
+    # Teardown: stop the loop and join the thread created during the test (if any)
+    # This ensures that each test (or the first client in it) starts fresh
+    test_loop = getattr(ToolboxSyncClient, "_ToolboxSyncClient__loop", None)
+    test_thread = getattr(ToolboxSyncClient, "_ToolboxSyncClient__thread", None)
+
+    if test_loop and test_loop.is_running():
+        test_loop.call_soon_threadsafe(test_loop.stop)
+    if test_thread and test_thread.is_alive():
+        # Wait for the run_forever loop to exit after stop()
+        test_thread.join(timeout=2)
+
+    # Restore original state (likely None if tests are isolated, but good practice)
+    ToolboxSyncClient._ToolboxSyncClient__loop = original_loop
+    ToolboxSyncClient._ToolboxSyncClient__thread = original_thread
```
This seems pretty heavy handed. Why not just recreate the client each time? Each client can have a smaller scope if we are worried about overlap.

anubhav756 replied:
Thanks for the suggestion! I've refactored the fixture setup to remove the `autouse`, and also the tests now explicitly request a new `sync_client` fixture.

kurtisvg reviewed (commented):
Just to confirm -- do we have a 1:1 mapping of sync tests to auth tests? I see some "Extra" sync tests. Should we move those into a Class so it's clear they are additive?

anubhav756 commented:
> Just to confirm -- do we have a 1:1 mapping of sync tests to auth tests? I see some "Extra" sync tests. Should we move those into a Class so it's clear they are additive?

Yes, there is a 1:1 mapping for core auth tests in `TestSyncAuth`.

Thanks for the suggestion! Updated the sync client tests

* Grouped the extra sync-specific lifecycle tests into a new `TestSyncClientLifecycle` class to make their additive nature clear.
* Moved the existing sync header tests into `TestSyncClientHeaders` for better structure.

kurtisvg commented on packages/toolbox-core/tests/test_sync_client.py:
```diff
@@ -0,0 +1,594 @@
+
+
+@pytest.fixture
+def sync_client(sync_client_environment, request):
+    """
+    Provides a ToolboxSyncClient instance within an isolated environment.
+    The client's underlying async session is automatically closed after the test.
+    The class-level loop/thread are managed by sync_client_environment.
+    """
+    # `sync_client_environment` has prepared the class state.
+    client = ToolboxSyncClient(TEST_BASE_URL)
+
+    def finalizer():
+        client.close()  # Closes the async_client's session.
+        # Loop/thread shutdown is handled by sync_client_environment's teardown.
+
+    request.addfinalizer(finalizer)
+    return client
```
Why cant this us yield + close after the yield?

anubhav756 replied:
Thanks for suggesting the simplification. 🙂

kurtisvg commented on packages/toolbox-core/tests/test_sync_client.py:
```diff
@@ -0,0 +1,594 @@
+    ToolboxSyncClient._ToolboxSyncClient__thread = None
+
+
+@pytest.fixture
```
If we scope this to "function" can we remove the "sync_client_envirtoment"?

kurtisvg replied:
https://docs.pytest.org/en/6.2.x/fixture.html#fixture-scopes

anubhav756 replied:
Simplifying fixtures sounds good, however, `sync_client_environment` ensures that these tests are properly isolated and reliable.

This is because `ToolboxSyncClient` uses a shared, class-level event loop/thread to be reused across the program. The `sync_client_environment` fixture specifically handles resetting these before and after each test.

Simply removing it would lead to state leakage (and potentially flaky tests), as `client.close()` alone doesn't currently perform this specific class-level reset.
